### PR TITLE
CFLAGS and LDFLAGS to find headers and libraries

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by jakirkham** [comment]
Excellent. Could you please bump the version to 2.1.0?

**Comment by jakirkham** [comment]
LGTM. Thanks @isuruf.

**Comment by isuruf** [comment]
@jakirkham, anything else to do?

**Comment by 183amir** [comment]
Do you have any argument behind this pull request?

**Comment by isuruf** [comment]
Headers are located in `$PREFIX/include`, but the compiler doesn't know that the headers are located there. Similarly for the libraries. This would fix that problem.

This conversation came up here, https://github.com/conda-forge/staged-recipes/pull/953 where a header (`gmp.h`) is not found, but can be found in `$PREFIX/include`

If this was added, specifying where the headers and libraries are located is not necessary. eg: [mpfr](https://github.com/conda-forge/mpfr-feedstock/blob/master/recipe/build.sh#L18), [mpc](https://github.com/conda-forge/mpc-feedstock/blob/master/recipe/build.sh#L18)

**Comment by 183amir** [comment]
That's what `PREFIX` is for but seems like those packages ignore that and provide `--with-gmp` arguments. I think adding this extra things here for just a few packages is going to clutter the `toolchain` package which makes maintenance and debugging very hard.

**Comment by isuruf** [comment]
`--prefix` is for installation directory and not for finding libraries and headers. If it were, then changing the prefix argument would change how the package is built and could trigger a new build and break the following rule from GNU configure.
> Running ‘make install’ with a different value of prefix from the one used to build the program should not recompile the program.

Btw, if it makes maintenance hard then I'm okay with not merging this.

File: recipe/activate.sh [hunk 1]
```diff
@@ -38,5 +38,8 @@ else
     exit 1
 fi
 
+export CFLAGS="${CFLAGS} -I${PREFIX}/include"
+export CXXFLAGS="${CXXFLAGS} -I${PREFIX}/include"
```
Comment (isuruf) [inline recipe/activate.sh]: This should have been just `CPPFLAGS` not `CXXFLAGS` and `CFLAGS`

**Comment by jakirkham** [comment]
> @jakirkham, anything else to do?

I try to let `toolchain` changes sit for a bit so that we get a broad spectrum of feedback. Though this seems fine to me.

**Comment by jakirkham** [comment]
> I think adding this extra things here for just a few packages is going to clutter the `toolchain` package which makes maintenance and debugging very hard.

@183amir, do you know of any cases where a change like this would cause problems? It seems pretty sane to me.

**Comment by 183amir** [comment]
Then that is the job of `PKG_CONFIG_PATH` which is set: http://conda.pydata.org/docs/building/environment-vars.html#environment-variables-set-during-the-build-process to help builds find libraries.
And when that doesn't happen and the program just look in `/usr/lib` you need more than what you are proposing to fix the problem sometimes. This pull works for you but will not work for `ffmpeg` trying to find `zlib`. For `ffmpeg` I had to add `-I$PREFIX` and `-L$PREFIX -lz`. See here: https://github.com/conda-forge/ffmpeg-feedstock/blob/master/recipe/build.sh#L8-10 I have used `pkg-config` because `zlib` had a `.pc` file but I could have just added `-I$PREFIX` and `-L$PREFIX -lz` to flags.

**Comment by ocefpaf** [comment]
> That's what `PREFIX` is for but seems like those packages ignore that and provide `--with-gmp` arguments. I think adding this extra things here for just a few packages is going to clutter the `toolchain` package which makes maintenance and debugging very hard.

I agree. IMO we should actually avoid a `toolchain` altogether.  RPM has a similar concept with the `optflags`, but for RPM the `optflags` do not play a role in the build nor install platform (or options) only in the mutil-platform package building (i.e. adding optional flags for a "matrix" of platforms). 

See: http://www.rpm.org/max-rpm/s1-rpm-multi-optflags.html

That makes a lot of sense to me because we get:

- explicit flags in every recipe
- better reproducibility (we could pin the `toolchain` for that but it sill leads to an extra package to debug) 
- avoid unknown reactions to the flags set in the `toolchain` like happened with `gdal`

**Comment by isuruf** [comment]
@183amir, this works for me, https://github.com/isuruf/ffmpeg-feedstock/commit/86d499640d9cd46ca6c8dd1244904f386ef3fe20

> avoid unknown reactions to the flags set in the toolchain like happened with gdal

I agree. Adding everything to toolchain would have unexpected consequences

**Comment by 183amir** [comment]
Your builds are failing: https://circleci.com/gh/isuruf/ffmpeg-feedstock but I guess that's not a `zlib` issue. Anyways, I think this is still not absolutely necessary and should not be added.

**Comment by scopatz** [comment]
Pinging this, @conda-forge/core

**Comment by jakirkham** [comment]
While this seems sensible, @183amir has a point that one can always add this later. So this doesn't need to be in the `toolchain` per se.

**Comment by scopatz** [comment]
Well it seems like if it doesn't harm things to add, then it would address some problems packages that don't rely on a build.sh script have.  Alternatively, if it is a pain to have some of these enabled everywhere, maybe we could have a toolchain-opt package that enables extras like this, and depends on toolchain.

**Comment by isuruf** [comment]
While I do know that we can always add this in a recipe builds script, but isn't it the whole point of toolchain to set these up for you? `BOOST_MATH_DISABLE_FLOAT128` is something added in toolchain feedstock and that's there to address problems of a very few feedstocks and this PR would be beneficial to a lot of feedstocks.

Conda already has some defaults. On windows `PREFIX/Library/bin` is added to path to make sure the dlls are found. This change is similar.

Also if you prefer to have flags explicitly in a recipe build script like @ocefpaf prefers, then it's just a matter of not using toolchain at all.

**Comment by scopatz** [comment]
Well, I view toolchain as *required* because it sets things where there would be binary incompatibilities or outright failures to build if they weren't set, like `BOOST_MATH_DISABLE_FLOAT128`.  We want to be consistent across the whole ecosystem with these, even if a package doesn't think it uses it.

Optional values could go elsewhere and either be used or not, like you said.  Right now optional values are explicitly in the recipe. One of the problems with this is if you use Python's build faciulities through distutils, etc. it may not be great at picking up header files in non-standard place and it is sort of a pain to add `INCLUDE_DIR` and friends to what would otherwise be a simple meta.yaml file.  This would be an optional thing to make recipes simpler

**Comment by 183amir** [comment]
Guys having had more experience with the `toolchain` package, I think it's a blessing especially when you are trying to manage hundreds of recipes. I am OK with this being added to the toolchain. But, we should always be careful because these kind of changes are **very** hard to debug. For example, the ffmpeg feedstock was [failing](https://github.com/conda-forge/ffmpeg-feedstock/issues/12) because of a variable named `SUBDIR` was set by `conda-build` and it took us a long time to figure out this was the cause.

**Comment by isuruf** [comment]
Ping. I'd like to get this and #17 in.

**Comment by scopatz** [comment]
@jakirkham @ocefpaf any objections? I am all for this.

**Comment by isuruf** [comment]
Packaging around 40 C/C++ packages for sagemath, I first took @183amir's advice and didn't add these variables in the recipes and added only for ones that failed. After half a dozen packages, I realized most of them failed or needed a configure option to find the package dir. So now, these 2 lines are in most of the C/C++ packages I maintain.

I'll make a `toolchain-opt` with changes from this PR and #17 if this doesn't fit in `toolchain`.

**Comment by isuruf** [comment]
See https://github.com/search?utf8=%E2%9C%93&q=org%3Aconda-forge+-I%24%7BPREFIX%7D%2Finclude&type=Code

**Comment by 183amir** [comment]
@isuruf I am sorry that you had go through that.
As I have said before, I'm ok with this being added to the toolchain. But I think a better effort before merging this in would be to pin toolchain on a specific version in the feedstocks so if some change in the toolchain breaks something, we would know.

**Comment by 183amir** [comment]
@isuruf pinning packages are done automatically. To add toolchain to the list, propose a merge request here: https://github.com/conda-forge/conda-forge.github.io/blob/master/scripts/pin_the_slow_way.py

**Comment by isuruf** [comment]
From what I remember, it's 20 maintenance jobs per day and assuming 10 per day for toolchain pinning, it'll take 40 days just to send out the PRs. Since this is a update pinnings run, CI will not be skipped.
I'm not that keen on stressing the CI like that.

Is it possible to run the script manually with `[skip ci]` in both commit message and PR title?

**Comment by 183amir** [comment]
I have no clue! better to ask some of the core members.

**Comment by isuruf** [comment]
Not needed anymore due to #23

**Comment by jakirkham** [comment]
I'd like to resurrect and merge this if it is ok.

**Comment by scopatz** [comment]
No complaints from me @jakirkham :)

**Comment by jakirkham** [comment]
AppVeyor failure will be fixed by PR ( https://github.com/conda-forge/toolchain-feedstock/pull/26 ).

**Comment by jakirkham** [comment]
Toggling for CIs.

**Comment by jakirkham** [comment]
Should we add `C_INCLUDE_PATH` as well? We are already effectively taking care of it here, but maybe it would address some odd cases to set it explicitly.

**Pull request merged**